### PR TITLE
 OS-8543 Want support for OFD locks in lx-brand

### PR DESCRIPTION
**Pull request by danmcd**
This was originally part of #467 , but because I'm taking it over, I need to reset this PR.

### CODE REVIEW

**Comment by danmcd** [comment]
@smokris ==> If you can, please try the updated fix in your environment?  I think I reproduced your failure-then-success case with a Debian 12 LX zone, but I would really appreciate if you could re-reproduce it to make sure my additional corner-case covering didn't break anything?

**Comment by danmcd** [comment]
Some additional testing notes are in the ticket.

File: usr/src/uts/common/brand/lx/syscall/lx_fcntl.c [hunk 1]
```diff
@@ -465,16 +502,43 @@ lx_fcntl_lock(int fd, int lx_cmd, void *arg)
...
 
+		if (cmd == F_OFD_GETLK || cmd == F_OFD_SETLK ||
+		    cmd == F_OFD_SETLKW) {
+			/*
+			 * TBD OFD-style locking is currently limited to
+			 * covering the entire file.
+			 */
+			if (bf.l_whence != 0 || bf.l_start != 0 ||
+			    bf.l_len != 0) {
+				/*
```
Comment (citrus-it) [inline usr/src/uts/common/brand/lx/syscall/lx_fcntl.c:513]: Could you consider adding an `lx_unsupported()` call here to make it easier to spot?
The dtrace probe doesn't hurt but I usually reach straight for `sdt:lx_brand::brand-lx-unsupported` when trying to debug something in lx.

**Review by citrus-it — COMMENTED** [review]
Thanks very much for doing this work.

**Comment by danmcd** [comment]
Updated test results on the OS-8543 ticket.

File: usr/src/uts/common/brand/lx/syscall/lx_fcntl.c (hunk 1 above)
Comment (danmcd) [reply usr/src/uts/common/brand/lx/syscall/lx_fcntl.c:514]: Done.

**Comment by danmcd** [comment]
@smokris can you please again try this branch in your environment?

**Comment by smokris** [comment]
Thanks for your work on this, @danmcd.

If I understand correctly, your changes (compared to my initial PR) are:

   - added comment and instrumentation regarding the lack of support for partial-file locking
   - added checking and struct conversion for 32-bit syscalls

I read through your changes (compared to my initial PR) and I think they make sense (to the limited extent that I understand the 32/64-bit compatibility layer).

I made a local PI build from your `OS-8543` branch and re-ran my test from https://github.com/TritonDataCenter/illumos-joyent/pull/467, and the test passes (that is, `systemd-sysusers` and `apt upgrade` both complete successfully).  LGTM.

**Comment by danmcd** [comment]
> I made a local PI build from your `OS-8543` branch and re-ran my test from #467, and the test passes (that is, `systemd-sysusers` and `apt upgrade` both complete successfully). LGTM.

Two favors:

1.) Please check a reboot of your zone to see whether or not systemd is running after your "apt upgrade" or "systemd-sysusers"?

2.) If your LGTM stands, please +1 this review so it's on the record. If, for permissions reasons, you cannot, I will note you as a reviewer in the commit message along with either you as author or you as portions-contributed-by.

**Comment by danmcd** [comment]
More testing in bug ticket.

**Review by citrus-it — APPROVED** [review]
Thanks for adding and testing the lx unsupported call.

**Comment by danmcd** [comment]
@TritonDataCenter/triton-engineering ==> Any objections for this landing in 20240531 ?

**Comment by danmcd** [comment]
Due to the details of how this was reviewed, I'm going to push from the outside:

```
commit 3c71165421e1b0fc36533828e5f2369cfff91f53 (HEAD -> master)
Author: Steve Mokris <steve@kosada.com>
Date:   Mon Apr 22 15:15:09 2024 -0400

    OS-8543 Want support for OFD locks in lx-brand
    Portions-contributed-by: Dan McDonald <danmcd@mnx.io>
    Reviewed by: Andy Fiddaman <illumos@fiddaman.net>
    Approved by: Brian Bennett <brbennett@mnx.io>
```

**Comment by danmcd** [comment]
Pushed.

**Comment by smokris** [comment]
Nice!  I'm a few minutes late, but here's my testing report:

While testing the below, I ran a dtrace probe for `sdt:lx_brand::brand-lx-unsupported`, and it caught two instances where LX processes attempted to do an unsupported ___partial-file___ OFD lock (details below).

## Testing systemd
> Please check a reboot of your zone to see whether or not systemd is running after your "apt upgrade" or "systemd-sysusers"?

After rebooting the zone, `systemd` is running as the init process as expected.
```console
# ps -q 1
  PID TTY          TIME CMD
    1 ?        00:00:36 systemd
# ls -l /proc/1/exe
lrwxrwxrwx 1 root root 0 May 28 19:21 /proc/1/exe -> /usr/lib/systemd/systemd
# systemctl status
● 3c03fcdf-9941-4bd9-8610-fd56e805b6db
    State: degraded
    Units: 181 loaded (incl. loaded aliases)
     Jobs: 0 queued
   Failed: 5 units
    Since: Tue 2024-05-28 19:21:46 UTC; 7min ago
  systemd: 254.5-1~bpo12+3
…
```
(The 5 failed units are an existing/separate problem unrelated to this PR.)

I searched the systemd source code and I think I found all the places it uses OFD locks:

- `systemd-sysusers` CLI — I tested it by running `systemd-sysusers ; echo $?`; it completed successfully
- `systemd-firstboot` CLI — I tested it by running `systemd-firstboot --prompt --reset ; echo $?`; it completed successfully
- [`DynamicUser=true`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#DynamicUser=) — can't test since it implies `PrivateTmp=true`, which requires Linux kernel namespaces, which LX doesn't currently support
- `systemd-nspawn` — can't test since it requires Linux kernel namespaces, which LX doesn't currently support
- `systemd-sysext` — can't test since it requires overlayfs, which LX doesn't currently support

So, I don't think this PR makes systemd support _worse_ than it was before.

## Testing other software packages
I searched for other software packages that use OFD locks:

- `bazel` — successfully launched and ran a build.  During that test run, the dtrace probe caught the following:
  ```
  1   5971 lx_unsupported:brand-lx-unsupported  28294              bazel-4 F_OFD_* fcntl being sent with one or more nonzero whence (0), start (0), or len (4096).\0
  ```
  Bazel successfully [fell back to using `F_SETLK`](https://github.com/bazelbuild/bazel/blob/463f80979127b39bb41c9afde5b5863914a7609d/src/main/cpp/blaze_util_posix.cc#L607-L626).

- `ceph` — can't test since it requires overlayfs

- `xapian` — successfully launched and created a database.  During that test run, the dtrace probe caught the following:
  ```
  1   5971 lx_unsupported:brand-lx-unsupported  21115           python3.11 F_OFD_* fcntl being sent with one or more nonzero whence (0), start (0), or len (1).\0
  ```
  Xapian successfully [fell back to using `O_CLOEXEC`](https://github.com/xapian/xapian/blob/9d1a8addfce61d07aa094c1c5b3efb905caa2556/xapian-core/backends/flint_lock.cc#L206-L216).

I also tested some other software packages which my team commonly uses on LX (I don't know offhand whether these packages use OFD locks), and performed some basic testing of each:

- `apache2` — successfully launched and performed a simple query
- `elasticsearch` — successfully launched and performed a simple query
- `facter` — successfully generated default facts
- `mariadb-server` — successfully launched and performed a simple query
- `nodejs` — successfully executed some statements
- `openjdk-17-jdk` — successfully executed some statements
- `openssh-server` — successfully launched and connected remotely
- `perl` — successfully executed some statements
- `php-fpm` — successfully launched and performed a simple query
- `postgresql` — successfully launched and performed a simple query
- `python` — successfully executed some statements
- `ruby` — successfully executed some statements

## Still LGTM
> If your LGTM stands, please +1 this review so it's on the record.

I haven't found any regressions introduced by this change.  My LGTM stands.

**Comment by danmcd** [comment]
THANK YOU SO MUCH for that extra test data, which is now copied into the OS-8543 ticket as well.  This will be in this week's 20240530 ("Spider-Man") release.

**Pull request closed without merging**